### DISCUSSION TITLE
About P2P direct connection without relay

### DISCUSSION THREAD

**Discussion by MrChenhtlss** [rustdesk/rustdesk#6689] [Q&A]
I am a user from China

Both the local version and the controlled version are 1.2.4

What I want to ask is why the success rate of P2P is so poor, most of the connections are through relay servers

But I use a third-party NAT hole punching tool, and the probability of successful TCP/UDP hole punching is very high.

**Comment by giox069** [comment]
Hi MrChenhtlss, I'm an user too (from Italy).
I'm fighting against TCP hole punching not working in RustDesk too. In my case I discovered that when the called rustdesk is behind a linux router which uses nftables masquerading, P2P fails and RustDesk fallbacks to relay server.
If the router is a pfSense there are no problems: P2P is setup correctly.
In this specific case (linux router with nftables) I have some tcpdump traces which can show the problem. But I cannot understand if the problem is the RustDesk Hole Punching algorithm which is not smart enough, or nftables is doing something wrong.
If someone with more experience on hole punching and/or nftables could help, I can send more info on this.

**Comment by MrChenhtlss** [comment]
Yes, I have also found such a problem. It should be that Rustdesk's
algorithm is not smart enough. There is a similar software called Todesk,
which has a very high success rate in P2P drilling

Giovanni Panozzo ***@***.***>于2023年12月27日 周三03:00写道：
